### PR TITLE
Feature: add user data to call

### PR DESCRIPTION
**Pull request by copiltembel**
Following [this discussion](https://github.com/baresip/baresip/discussions/1901), I am making this PR which allows a user to add some **custom data** to an **outgoing** call. This custom user data can be propagated back whenever an event occurs on the call.

Usage examples:
1. Specifying the ua
`/dial +123456789 0 userdata=some_custom_data`
2. Not specifying the ua:
`/dial +123456789 userdata=some_custom_data`

The userdata parameter is **optional**. All commands that don't use it will behave just like before.

The userdata will be linked to the call and therefore can be accessed wherever the call is available. If present, it is also serialized in the `event_encode_dict` function.

I've tested this only with the `/dial` verb from the command line as well as from the mqtt module:
`mosquitto_pub -t /baresip/command/ -m '{"command":"dial","params":"+123456789 userdata=my_custom_data","token":"123"}'`

And here are some examples of the call status updates received on an mqtt client. Note the `userdata`:

```
/baresip/event {"type":"CALL_OUTGOING","class":"call","accountaor":"sip:+0987654321@somedomain.com","direction":"outgoing","peeruri":"sip:+123456789@somedomain.com;transport=tls","id":"bd55327bec82600f","remoteaudiodir":"inactive","remotevideodir":"inactive","audiodir":"inactive","videodir":"inactive","userdata":"my_custom_data","param":"sip:+123456789@somedomain.com;transport=tls"}
/baresip/event {"type":"CALL_LOCAL_SDP","class":"other","accountaor":"sip:+0987654321@somedomain.com","direction":"outgoing","peeruri":"sip:+123456789@somedomain.com;transport=tls","id":"bd55327bec82600f","remoteaudiodir":"inactive","remotevideodir":"inactive","audiodir":"inactive","videodir":"inactive","userdata":"my_custom_data","param":"offer"}
/baresip/event {"type":"CALL_RINGING","class":"call","accountaor":"sip:+0987654321@somedomain.com","direction":"outgoing","peeruri":"sip:+123456789@somedomain.com;transport=tls","id":"bd55327bec82600f","remoteaudiodir":"inactive","remotevideodir":"inactive","audiodir":"inactive","videodir":"inactive","userdata":"my_custom_data","param":"sip:+123456789@somedomain.com;transport=tls"}
/baresip/event {"type":"CALL_CLOSED","class":"call","accountaor":"sip:+0987654321@somedomain.com","direction":"outgoing","peeruri":"sip:+123456789@somedomain.com;transport=tls","id":"bd55327bec82600f","remoteaudiodir":"inactive","remotevideodir":"inactive","audiodir":"inactive","videodir":"inactive","userdata":"my_custom_data","param":"480 Temporarily Unavailable"}
```

### CODE REVIEW

**Comment by alfredh** [comment]
thanks!

there are some build errors:

```
/home/runner/work/baresip/baresip/src/event.c: In function ‘event_encode_dict’:
/home/runner/work/baresip/baresip/src/event.c:211:21: error: initialization discards ‘const’ qualifier from pointer target type [-Werror=discarded-qualifiers]
  211 |   char *user_data = call_user_data(call);
      |                     ^~~~~~~~~~~~~~
```

could you please take a look ?

**Comment by copiltembel** [comment]
Thanks, done! Somehow I didn't get that error when building locally.

**Comment by copiltembel** [comment]
Hm, that's fishy! Thanks for the run, I'll have another look!

**Comment by copiltembel** [comment]
I have fixed the tests and then ran `make test`. But I get this error:

```
/usr/bin/ld: ./libbaresip.a(module.o): in function `lookup_static_module':
/home/tdr/src/github/baresip/baresip/src/module.c:44: undefined reference to `mod_table'
collect2: error: ld returned 1 exit status
make: *** [Makefile:339: selftest] Error 1
```

**Comment by copiltembel** [comment]
Btw, it would be good to add somthing like

* The code must pass "make test"

to the CONTRIBUTING.md

(if that loader issue can be fixed also locally)

File: include/baresip.h [hunk 1]
```diff
@@ -855,7 +856,7 @@ int  ua_disable_autoanswer(struct ua *ua, enum answer_method met);
 int  ua_call_alloc(struct call **callp, struct ua *ua,
 		   enum vidmode vidmode, const struct sip_msg *msg,
 		   struct call *xcall, const char *local_uri,
-		   bool use_rtp);
+		   bool use_rtp, const char *user_data);
```
Comment (cspiel1) [inline include/baresip.h:859]: Can't you use a setter function for the user_data?
You had already one, why you removed it?

File: modules/menu/static_menu.c [hunk 2]
```diff
@@ -556,7 +557,15 @@ static int dial_handler(struct re_printf *pf, void *arg)
...
 
 	re_hprintf(pf, "call uri: %s\n", uri);
-	err = ua_connect(ua, &call, NULL, uri, VIDMODE_ON);
+
+	const char ud_sentinel[] = "userdata=";
+	char *ud_pos = strstr(carg->prm, ud_sentinel);
+	char *user_data = NULL;
+	if (ud_pos != NULL) {
+		user_data = ud_pos + strlen(ud_sentinel);
+	}
```
Comment (cspiel1) [inline modules/menu/static_menu.c:566]: In baresip `re_regex` is often used for parsing. This is much more powerful and should simplify the code. Maybe we should extend the function `menu_ua_carg()` to parse also `userdata=xxx`.

File: src/call.c [hunk 3]
```diff
@@ -869,7 +871,7 @@ int call_alloc(struct call **callp, const struct config *cfg, struct list *lst,
 	       struct account *acc, struct ua *ua, const struct call_prm *prm,
 	       const struct sip_msg *msg, struct call *xcall,
 	       struct dnsc *dnsc,
-	       call_event_h *eh, void *arg)
+	       call_event_h *eh, void *arg, const char *user_data)
```
Comment (cspiel1) [inline src/call.c:874]: Is it possible to use a setter function instead of extending the parameter list?

File: modules/menu/static_menu.c (hunk 2 above)
Comment (copiltembel) [reply modules/menu/static_menu.c:566]: I've spent some time trying to find a regex for it. The best that I could come up with is this: 

`([^ ]+) ?([1234567890]+)? ?(userdata=[^ ]+)?`

which I tested here https://regex101.com with the following inputs:

```
+123456789
+123456789 0
+123456789 0 userdata=somedata
+123456789 userdata=somedata
```
Unfortunately, the `re_regex` function doesn't seem to behave the same.

File: include/baresip.h (hunk 1 above)
Comment (copiltembel) [reply include/baresip.h:859]: Looking through the code it felt more natural to me to set the user_data when the call object is created.

File: src/call.c (hunk 3 above)
Comment (copiltembel) [reply src/call.c]: OK, I've added a setter.

File: include/baresip.h (hunk 1 above)
Comment (copiltembel) [reply include/baresip.h]: I've changed the code to use a setter instead.

File: modules/menu/menu.c [hunk 4]
```diff
@@ -681,10 +681,14 @@ static void ua_event_handler(struct ua *ua, enum ua_event ev,
 		     call_id(call), prm);
 
 		err = ua_call_alloc(&call2, ua, VIDMODE_ON, NULL, call,
-				    call_localuri(call), true);
+		    call_localuri(call), true);
```
Comment (sreimers) [inline modules/menu/menu.c:684]: unrelated change

File: test/call.c [hunk 5]
```diff
@@ -340,9 +340,6 @@ static void event_handler(struct ua *ua, enum ua_event ev,
 		if (!err) {
 			struct pl pl;
 
-			/* unsure about this one, would probably be best
-			to allow the user to set something new
-			on user_data */
 			call_set_user_data(call2, call_user_data(call));
```
Comment (copiltembel) [inline test/call.c:343]: unsure about this one, would probably be best to allow the user to set something new on user_data

File: src/call.c [hunk 6]
```diff
@@ -2967,3 +2969,33 @@ bool call_supported(struct call *call, uint16_t tags)
...
+ * @param user_data User data to be set
+ * @return int
+ */
+
+int call_set_user_data(struct call *call, const char *user_data)
+{
+	if (!call)
+		return EINVAL;
+
+	str_dup(&call->user_data, user_data);
```
Comment (alfredh) [inline src/call.c:2998]: the previous user_data should be freed with mem_deref first

`call->user_data = mem_deref(call->user_data`

File: src/call.c [hunk 7]
```diff
@@ -2967,3 +2969,33 @@ bool call_supported(struct call *call, uint16_t tags)
...
+ * @param call Call object
+ * @param user_data User data to be set
+ * @return int
+ */
+
+int call_set_user_data(struct call *call, const char *user_data)
+{
+	if (!call)
+		return EINVAL;
+
```
Comment (alfredh) [inline src/call.c:2997]: please check return value of `str_dup`

File: src/call.c [hunk 8]
```diff
@@ -2967,3 +2969,33 @@ bool call_supported(struct call *call, uint16_t tags)
...
+	str_dup(&call->user_data, user_data);
+
+	return 0;
+}
```
Comment (alfredh) [inline src/call.c:3001]: files must contain a newline
